README.md
# csc207-2023fa-associative-array-tests

Tests for our AssociativeArray class (Mini-Project 4).

Suggested strategy:

* Clone this repo.
* Make a soft link to your `structures` directory in the `src` directory.  
  For example,

        ln -s /home/username/CSC207/MP4/src/structures structures

* Open this repository in VSCode.
* Cross your fingers.
* Run the tests.


CHANGES.md
Hello! changes I made was:
ensure that all my tests pass.
Additionally, I added some more documentation,and fixed the incorrect package issue by removing the unnecesary package declaration.
The only tests that fail are sam's, which are designed to fail purposely by him.


KVPair.java

/**
 * An easy way to store key/value pairs. We assume that other
 * classes will access fields directly.
 */
class KVPair<K, V> {
  // +--------+------------------------------------------------------
  // | Fields |
  // +--------+

  /**
   * The key.
   */
  K key;

  /**
   * The value.
   */
  V value;

  // +--------------+------------------------------------------------
  // | Constructors |
  // +--------------+

  /**
   * Create an empty key/value pair.
   */
  KVPair() {
    this(null, null);
  } // KVPair()

  /**
   * Create a new key/value pair.
   */
  KVPair(K key, V value) {
    this.key = key;
    this.value = value;
  } // KVPair(K,V)

  // +------------------+--------------------------------------------
  // | Standard methods |
  // +------------------+

  public KVPair<K, V> clone() {
    return new KVPair<K, V>(this.key, this.value);
  } // clone()

  public String toString() {
    return "{ " + this.key.toString() + " : " + this.value.toString() + " }";
  } // toString()
} // class KVPair


AssociativeArray.java

import static java.lang.reflect.Array.newInstance;

import java.util.Arrays;

/**
 * A basic implementation of Associative Arrays with keys of type K
 * and values of type V. Associative Arrays store key/value pairs
 * and permit you to look up values by key.
 *
 * @author John Miller
 * @author Samuel A. Rebelsky
 */
public class AssociativeArray<K, V> {
  // +-----------+---------------------------------------------------
  // | Constants |
  // +-----------+

  /**
   * The default capacity of the initial array.
   */
  static final int DEFAULT_CAPACITY = 16;

  // +--------+------------------------------------------------------
  // | Fields |
  // +--------+

  /**
   * The size of the associative array (the number of key/value pairs).
   */
  int size;

  /**
   * The array of key/value pairs.
   */
  KVPair<K, V> pairs[];

  // +--------------+------------------------------------------------
  // | Constructors |
  // +--------------+

  /**
   * Create a new, empty associative array.
   */
  @SuppressWarnings({ "unchecked" })
  public AssociativeArray() {
    // Creating new arrays is sometimes a PITN.
    this.pairs = (KVPair<K, V>[]) newInstance((new KVPair<K, V>()).getClass(),
        DEFAULT_CAPACITY);
    this.size = 0;
  } // AssociativeArray()

  // +------------------+--------------------------------------------
  // | Standard Methods |
  // +------------------+

  /**
   * Create a copy of this AssociativeArray.
   */
  public AssociativeArray<K, V> clone() {
    AssociativeArray<K, V> newArray = new AssociativeArray<K, V>();
    for (int i = 0; i < this.pairs.length; i++) {
      if (this.pairs[i] != null) {
        newArray.set(this.pairs[i].key, this.pairs[i].value);
      }
    }
    newArray.size = newArray.size();
    return newArray;
  } // clone()

  /**
   * Convert the array to a string.
   */
  public String toString() {
    String s = "{";// initial empty String
    for (int i = 0; i < this.pairs.length; i++) {
      if (this.pairs[i] != null) {
        s = s + "key" + i + ":";
        s = s + this.pairs[i].key.toString();
        s = s + ",";// providing a comma betwen the key and the value
        s = s + "value" + i + ": ";
        s = s + this.pairs[i].value.toString();
        s = s + " ";// providing spacing between pairs
      } else {
        s = s + "null, ";
      }
    }
    s = s + "}";
    return s;
  } // toString()

  // +----------------+----------------------------------------------
  // | Public Methods |
  // +----------------+

  /**
   * Set the value associated with key to value. Future calls to
   * get(key) will return value.
   * 
   * @pre: key is a valid K, value is a valid V
   * 
   * @Post: returns nothing but updates the AssociativeArray with the updated pair
   */
  @SuppressWarnings({ "unchecked" })
  public void set(K key, V value) {
    Boolean foundValue = false;
    int index = 0;
    int OpenIndex = 0;// find the first null index of the array for a potential storage of the key and
                      // value if not found initially
    boolean foundOpen = false;// find the first null index of the array for a potential storage of the key and
                              // value if not found initially
    while (foundValue == false && index < this.pairs.length) {
      if (this.pairs[index] == null) {// if we are at a null index
        if (foundOpen == false)// if this is the first null index that we have seen
        {
          foundOpen = true;
          OpenIndex = index;
        }
      } else if (this.pairs[index].key.equals(key) == true) {// if the pair at the index we are at has the key
        this.pairs[index].value = value;
        foundValue = true;
      }
      index++;
    }
    if (foundValue == false && foundOpen == true) {// this means we can't find the key int the array, but there is an
                                                   // opening in our array
      KVPair<K, V> newPair = new KVPair<K, V>(key, value);
      this.pairs[OpenIndex] = newPair;
      size++;
    } else if (foundValue == false && foundOpen == false) {// this means we couldn't find the key, and have to expand
                                                           // the array
      this.expand();
      KVPair<K, V> newPair = new KVPair<K, V>(key, value);
      this.pairs[this.size] = newPair;
      this.size++;
    }
  } // set(K,V)

  /**
   * Get the value associated with key.
   *
   * @throws KeyNotFoundException
   *                              when the key does not appear in the associative
   *                              array.
   *                              will give an error message if the key is null,
   *                              and if the key doesn't appear.
   * 
   * @pre: this is a valid AssociativeArray
   * 
   * @post: returns V value, or error message
   */
  public V get(K key) throws KeyNotFoundException {

    if (key == null) {
      throw new Error("the key is null, therefore it couldn't be stored");
    }

    try {
      int i = this.find(key);
      return this.pairs[i].value;
    } catch (KeyNotFoundException K) {
      throw new Error("there is no value corresponding to this key in the associate array\n");
    }
  } // get(K)

  /**
   * Determine if key appears in the associative array.
   * 
   * @pre: this is a valid AssociativeArray
   * 
   * @Post: returns true is the Array has K key
   * 
   */
  public boolean hasKey(K key) {
    if (key == null) {
      return false;// return false if the key is null
    }
    for (int i = 0; i < this.size(); i++) {
      if (this.pairs[i] != null) {
        if (this.pairs[i].key.equals(key) == true) {
          return true;
        }
      }
    }
    return false;// if the code in the for loop doesn't execute, return false
  } // hasKey(K)

  /**
   * Remove the key/value pair associated with a key. Future calls
   * to get(key) will throw an exception. If the key does not appear
   * in the associative array, does nothing.
   * 
   * @pre: this is a valid AssociativeArray
   * 
   * @Post: removes the key, or throws an error message
   * 
   */
  public void remove(K key) {
    try {
      int index = this.find(key);
      this.pairs[index] = null;
      this.size = this.size();
    } catch (KeyNotFoundException KNFE) {
      throw new Error("the key was not found");
    }
  }// remove(K)

  /**
   * Determine how many values are in the associative array.
   * 
   * @pre: this is a valid associative array
   * 
   * @post: returns an integer representing how many values are in it.
   */
  public int size() {
    int sizeCount = 0;
    for (int i = 0; i < this.pairs.length; i++) {
      if (this.pairs[i] != null) {// if we are not at a null index
        sizeCount++;
      }
    }
    return sizeCount;
  } // size()

  // +-----------------+---------------------------------------------
  // | Private Methods |
  // +-----------------+

  /**
   * Expand the underlying array.
   */
  public void expand() {
    this.pairs = java.util.Arrays.copyOf(this.pairs, this.pairs.length * 2);
  } // expand()

  /**
   * Find the index of the first entry in `pairs` that contains key.
   * If no such entry is found, throws an exception.
   */
  public int find(K key) throws KeyNotFoundException {
    for (int i = 0; i < this.pairs.length; i++) {
      if (this.pairs[i] != null) {
        if (this.pairs[i].key.equals(key) == true) {
          return i;
        }
      }
    }
    throw new KeyNotFoundException("The key couldn't be found by the find method\n");
  }// find(K)

} // class AssociativeArray


AssociativeArrayTestsMain.java

/**
 * A quick hack to make it easier to debug with tests.
 */

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

public class AssociativeArrayTestsMain {
  /**
   * A test of cloning.
   */
  @Test
  public void samuelRebelskyTest01() {
    // Build an array
    AssociativeArray<String, String> arr = new AssociativeArray<String, String>();
    arr.set("A", "Apple");
    try {
      assertEquals("Apple", arr.get("A"));
    } catch (Exception e) {
      fail("Original array does not contain expected value");
    }
    // Make a copy
    AssociativeArray<String, String> arr2 = arr.clone();
    // Make sure it contains the appropriate value
    try {
      assertEquals("Apple", arr2.get("A"));
    } catch (Exception e) {
      fail("Clone does not contain original value");
    } // try/catch
    // Change the original array
    arr.set("A", "aardvark");
    // Make sure we haven't changed the clone.
    try {
      assertEquals("Apple", arr2.get("A"));
    } catch (Exception e) {
      fail("Change to original changes clone.");
    }
    // Change the clone
    arr2.set("A", "Ant");
    // And look for values
    try {
      assertEquals("Ant", arr2.get("A"));
    } catch (Exception e) {
      fail("Cannot change clone");
    }
    try {
      assertEquals("aardvark", arr.get("A"));
    } catch (Exception e) {
      fail("Change to clone changes original");
    }
  } // samuelRebelskyTest01()

  /**
   * Can we successfully add a bunch of values? (Checks array expansion.)
   */
  @Test
  public void samuelRebelskyTest02() {
    AssociativeArray<Integer, Integer> arr = new AssociativeArray<Integer, Integer>();
    // Add a bunch of values
    for (int i = 10; i < 50; i++) {
      arr.set(i, i * i);
    } // for
    try {
      for (int i = 49; i >= 10; i--) {
        assertEquals(i * i, arr.get(i));
      }
    } catch (Exception e) {
      fail("Exception in call to get");
    }
  } // samuelRebelskyTest02()

  /**
   * Do we get exceptions when grabbing a deleted value from the array?
   */
  @Test
  public void samuelRebelskyTest03() {
    AssociativeArray<String, String> arr = new AssociativeArray<String, String>();
    // Add an element to the array
    arr.set("A", "Apple");
    // Make sure that it's there.
    try {
      assertEquals("Apple", arr.get("A"));
    } catch (KeyNotFoundException e) {
      fail("Could not set A to Apple");
    }
    // Remove it.
    arr.remove("A");
    // Make sure it's no longer there.
    try {
      // The following line should throw an exception
      arr.get("A");
      fail("Did not throw an exception");
    } catch (KeyNotFoundException e) {
      // Do nothing
    }
  } // samuelRebelskyTest03
} // AssociativeArrayTestsMain


ReportingAssociativeArray.java
import java.io.PrintWriter;

/**
 * An extension of the AssociativeArray class that reports each
 * step it takes.
 *
 * @author Samuel A. Rebelsky
 */
public class ReportingAssociativeArray<K, V> extends AssociativeArray<K, V> {

    // +--------+------------------------------------------------------
    // | Fields |
    // +--------+

    /**
     * The "name" of this associative array (at least as reported to the user).
     */
    String name;

    /**
     * Where we print reports.
     */
    PrintWriter pen;

    // +--------------+------------------------------------------------
    // | Constructors |
    // +--------------+

    /**
     * Create a new ReportingAssociativeArray named `name` that
     * prints reports of all actions using `pen`.
     */
    public ReportingAssociativeArray(String name, PrintWriter pen) {
        super();
        this.name = name;
        this.pen = pen;
    } // ReportingAssociativeArray

    // +---------+-----------------------------------------------------
    // | Methods |
    // +---------+

    /**
     * Set a value.
     */
    public void set(K key, V value) {
        pen.println(name + ".set(" + key + ", " + value + ")");
        super.set(key, value);
        pen.println("  " + name + ": " + this.toString());
    } // set (K,V)

    /**
     * Get the value associated with key.
     *
     * @throws KeyNotFoundException
     *                              when the key does not appear in the associative
     *                              array.
     */
    public V get(K key) throws KeyNotFoundException {
        pen.print(name + ".get(" + key + ") = ");
        pen.flush();
        try {
            V value = super.get(key);
            pen.println(value);
            return value;
        } catch (KeyNotFoundException knfe) {
            pen.println("** ERROR **");
            throw knfe;
        } // try/catch
    } // get(K)

    /**
     * Determine if key appears in the associative array.
     */
    public boolean hasKey(K key) {
        pen.print(name + ".hasKey(" + key + ") = ");
        pen.flush();
        boolean result = super.hasKey(key);
        pen.println(result);
        return result;
    } // hasKey(K)

    /**
     * Remove the key/value pair associated with a key. Future calls
     * to get(key) will throw an exception. If the key does not appear
     * in the associative array, does nothing.
     */
    public void remove(K key) {
        pen.println(name + ".remove(" + key + ")");
        super.remove(key);
        pen.println("  " + name + ": " + this.toString());
    } // remove(K)

    /**
     * Determine how many values are in the associative array.
     */
    public int size() {
        pen.print(name + ".size() = ");
        int size = super.size(); // No, not your soda
        pen.println(size);
        return size;
    } // size()
} // ReportingAssociativeArray

AssociativeArrayExperiments.java
import java.io.PrintWriter;
import java.math.BigInteger;

/**
 * Experiments with our AssociativeArray class.
 *
 * @author John Miller
 * @author Samuel A. Rebelsky
 */
public class AssociativeArrayExperiments {

    // +------+--------------------------------------------------------
    // | Main |
    // +------+

    /**
     * Run the experiments.
     */
    public static void main(String[] args) {
        PrintWriter pen = new PrintWriter(System.out, true);

        divider(pen);
        expreimentStringsToStrings(pen);
        divider(pen);
        experimentBigIntToBigInt(pen);
        divider(pen);
    } // main(String[])

    // +-------------+-------------------------------------------------
    // | Experiments |
    // +-------------+

    /**
     * Our first experiment: Associative arrays with strings as both keys
     * and values.
     */
    public static void expreimentStringsToStrings(PrintWriter pen) {
        AssociativeArray<String, String> s2s = new ReportingAssociativeArray<String, String>("s2s", pen);
        s2s.size();
        s2s.set("a", "apple");
        s2s.set("A", "aardvark");
        s2s.size();
        s2s.hasKey("a");
        s2s.hasKey("A");
        try {
            s2s.get("a");
        } catch (Exception e) {
        }
        try {
            s2s.get("A");
        } catch (Exception e) {
        }
        // s2s.remove("a");
        s2s.size();
        try {
            s2s.get("a");
        } catch (Exception e) {
        }
        try {
            s2s.get("A");
        } catch (Exception e) {
        }
        // s2s.remove("aardvark");
        s2s.size();
        try {
            s2s.get("a");
        } catch (Exception e) {
        }
        try {
            s2s.get("A");
        } catch (Exception e) {
        }
    } // expreimentStringsToStrings

    /**
     * Our second experiment: Associative arrays with big integers as
     * keys and values.
     */
    public static void experimentBigIntToBigInt(PrintWriter pen) {
        AssociativeArray<BigInteger, BigInteger> b2b = new ReportingAssociativeArray<BigInteger, BigInteger>("b2b",
                pen);

        // Set some values
        for (int i = 0; i < 11; i++) {
            b2b.set(BigInteger.valueOf(i), BigInteger.valueOf(i * i));
        } // for

        // Then get them
        for (int i = 0; i < 11; i++) {
            try {
                b2b.get(BigInteger.valueOf(i));
            } catch (Exception e) {
            }
        } // for

        // Then remove some of them
        for (int i = 1; i < 11; i += 2) {
            b2b.remove(BigInteger.valueOf(i));
        } // for

        // Then see what happens when we get them
        for (int i = 0; i < 11; i++) {
            try {
                b2b.get(BigInteger.valueOf(i));
            } catch (Exception e) {
            }
        } // for

        // Then reset or set some values
        for (int i = 0; i < 11; i += 3) {
            b2b.set(BigInteger.valueOf(i), BigInteger.valueOf(i + 10));
        } // for

        // Then see what happens when we get them
        for (int i = 0; i < 11; i++) {
            try {
                b2b.get(BigInteger.valueOf(i));
            } catch (Exception e) {
            }
        } // for
    } // experimentBigIntToBigInt

    // +---------+-----------------------------------------------------
    // | Helpers |
    // +---------+

    /**
     * Print a divider.
     */
    public static void divider(PrintWriter pen) {
        pen.println();
        pen.println("------------------------------------------------");
        pen.println();
    } // divider(PrintWriter)

} // AssociativeArrayExperiments

AssociativeArrayTests.java
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

import java.math.BigInteger;

import org.junit.jupiter.api.Test;

/**
 * Tests of the AssociativeArray class.
 *
 * @author CSC-207 2023Fa
 */
public class AssociativeArrayTests {

  // +---------------------+-----------------------------------------
  // | Tests by Sam Bigham |
  // +---------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Micah Cantor |
  // +-----------------------+

  // +------------------------+--------------------------------------
  // | Tests by Reed Colloton |
  // +------------------------+

  // +------------------+--------------------------------------------
  // | Tests by Pom Dao |
  // +------------------+

  // +--------------------------+------------------------------------
  // | Tests by Joshua Delarosa |
  // +--------------------------+

  // +-------------------+-------------------------------------------
  // | Tests by Jinny Eo |
  // +-------------------+

  // +---------------------------+-----------------------------------
  // | Tests by Kevin Fitzgerald |
  // +---------------------------+

  // +---------------------+-----------------------------------------
  // | Tests by Joyce Gill |
  // +---------------------+

  // +--------------------+------------------------------------------
  // | Tests by Che Glenn |
  // +--------------------+

  // +-------------------------+-------------------------------------
  // | Tests by Kevin Johanson |
  // +-------------------------+

  // +----------------------+----------------------------------------
  // | Tests by Chloe Kelly |
  // +----------------------+

  // +--------------------+------------------------------------------
  // | Tests by Hyeon Kim |
  // +--------------------+

  // +---------------------+-----------------------------------------
  // | Tests by Julian Kim |
  // +---------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Jason Kunkel |
  // +-----------------------+

  // +---------------------+-----------------------------------------
  // | Tests by Wenfei Lin |
  // +---------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Noah Mendola |
  // +-----------------------+

  // +----------------------+----------------------------------------
  // | Tests by John Miller |
  // +----------------------+

  /* this tests the clone method */
  @Test
  public void johnMillerTest01() {
    // Build an array
    AssociativeArray<String, String> array = new AssociativeArray<String, String>();
    array.set("p", "platypus");// set a key/value pair in the AssociativeArray
    AssociativeArray<String, String> aClone = array.clone();
    try {
      assertEquals("platypus", aClone.get("p"));// checking that the new array contains the value from the initial array
    } catch (Exception e) {
      fail("Clone was not properly copied");
    }
  }

  /* this tests the remove method */
  @Test
  public void johnMillerTest02() {
    // Build an array
    AssociativeArray<String, String> array = new AssociativeArray<String, String>();
    array.set("p", "platypus");// set a key/value pair in the AssociativeArray
    array.set("g", "Giraffe");
    array.remove("p");// remove platypus from the array
    try {
      assertEquals(false, array.hasKey("p"));// checking that the p key was removed
    } catch (Exception e) {
      fail("The 'p' key was not properly removed\n");
    }
  }

  /* this tests the ability of get to find and overwrite values */
  @Test
  public void johnMillerTest03() {
    // Build an array
    AssociativeArray<BigInteger, String> array = new AssociativeArray<BigInteger, String>();
    array.set(BigInteger.valueOf(8), "platypus");// set a key/value pair in the AssociativeArray
    array.set(BigInteger.valueOf(15), "Giraffe");// set a key/value pair in the AssociativeArray
    array.set(BigInteger.valueOf(8), "dog");// set a key/value pair in the AssociativeArray
    try {
      assertEquals("dog", array.get(BigInteger.valueOf(8)));// checking that the key was overwritten
    } catch (Exception e) {
      fail("The value at key BigInt 8 wasn't properly overwritten\n");
    }
  }

  /*
   * this is an edge case for getting the size of an empty array. it should be
   * zero
   */
  @Test
  public void johnMillerEdge01() {
    AssociativeArray<String, String> array = new AssociativeArray<String, String>();
    assertEquals(array.size(), 0);
  }

  /*
   * this is an edge case for setting a null value. it should return the null
   * value at the key
   */
  @Test
  public void johnMillerEdge02() {
    AssociativeArray<BigInteger, String> array = new AssociativeArray<BigInteger, String>();
    array.set(BigInteger.valueOf(8), null);// setting the 8,null pair
    try {
      assertEquals(null, array.get(BigInteger.valueOf(8)));// asserting we get null at 8
    } catch (KeyNotFoundException e) {
      e.printStackTrace();
    }
  }

  // +-----------------------+---------------------------------------
  // | Tests by Albert Okine |
  // +-----------------------+

  // +---------------------+-----------------------------------------
  // | Tests by Alma Ordaz |
  // +---------------------+

  // +-----------------------------+---------------------------------
  // | Tests by Samuel A. Rebelsky |
  // +-----------------------------+

  /**
   * A test of cloning.
   */
  @Test
  public void samuelRebelskyTest01() {
    // Build an array
    AssociativeArray<String, String> arr = new AssociativeArray<String, String>();
    arr.set("A", "Apple");
    try {
      assertEquals("Apple", arr.get("A"));
    } catch (Exception e) {
      fail("Original array does not contain expected value");
    }
    // Make a copy
    AssociativeArray<String, String> arr2 = arr.clone();
    // Make sure it contains the appropriate value
    try {
      assertEquals("Apple", arr2.get("A"));
    } catch (Exception e) {
      fail("Clone does not contain original value");
    } // try/catch
    // Change the original array
    arr.set("A", "aardvark");
    // Make sure we haven't changed the clone.
    try {
      assertEquals("Apple", arr2.get("A"));
    } catch (Exception e) {
      fail("Change to original changes clone.");
    }
    // Change the clone
    arr2.set("A", "Ant");
    // And look for values
    try {
      assertEquals("Ant", arr2.get("A"));
    } catch (Exception e) {
      fail("Cannot change clone");
    }
    try {
      assertEquals("aardvark", arr.get("A"));
    } catch (Exception e) {
      fail("Change to clone changes original");
    }
  } // samuelRebelskyTest01()

  /**
   * Can we successfully add a bunch of values? (Checks array expansion.)
   */
  @Test
  public void samuelRebelskyTest02() {
    AssociativeArray<Integer, Integer> arr = new AssociativeArray<Integer, Integer>();
    // Add a bunch of values
    for (int i = 10; i < 50; i++) {
      arr.set(i, i * i);
    } // for
    try {
      for (int i = 49; i >= 10; i--) {
        assertEquals(i * i, arr.get(i));
      }
    } catch (Exception e) {
      fail("Exception in call to get");
    }
  } // samuelRebelskyTest02()

  /**
   * Do we get exceptions when grabbing a deleted value from the array?
   */
  @Test
  public void samuelRebelskyTest03() {
    AssociativeArray<String, String> arr = new AssociativeArray<String, String>();
    // Add an element to the array
    arr.set("A", "Apple");
    // Make sure that it's there.
    try {
      assertEquals("Apple", arr.get("A"));
    } catch (KeyNotFoundException e) {
      fail("Could not set A to Apple");
    }
    // Remove it.
    arr.remove("A");
    // Make sure it's no longer there.
    try {
      // The following line should throw an exception
      arr.get("A");
      fail("Did not throw an exception");
    } catch (KeyNotFoundException e) {
      // Do nothing
    }
  } // samuelRebelskyTest03

  /**
   * Do we get exceptions when grabbing a value from the empty array.
   */
  @Test
  public void samuelRebelskyEdge01() {
    AssociativeArray<String, String> arr = new AssociativeArray<String, String>();
    try {
      // The following line should throw an exception
      arr.get("A");
      fail("Did not throw an exception");
    } catch (KeyNotFoundException e) {
      // Do nothing
    }
  } // samuelRebelskyEdge01

  // +--------------------------+------------------------------------
  // | Tests by Maria Rodriguez |
  // +--------------------------+

  // +-----------------------------+---------------------------------
  // | Tests by Gabriela Roznawska |
  // +-----------------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Shuta Shibue |
  // +-----------------------+

  // +----------------------+----------------------------------------
  // | Tests by Madel Sibal |
  // +----------------------+

  // +------------------------------+--------------------------------
  // | Tests by Livia Stein Freitas |
  // +------------------------------+

  // +------------------------+--------------------------------------
  // | Tests by Tyrell Taylor |
  // +------------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Audrey Trinh |
  // +-----------------------+

  // +---------------------+-----------------------------------------
  // | Tests by Rene Urias |
  // +---------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Christina Vu |
  // +-----------------------+

  // +-----------------------+---------------------------------------
  // | Tests by Luke Walters |
  // +-----------------------+

  // +------------------------+--------------------------------------
  // | Tests by Jonathan Wang |
  // +------------------------+

  // +-------------------+-------------------------------------------
  // | Tests by Lydia Ye |
  // +-------------------+

} // class AssociativeArrayTests
